libkbfs/prefetcher.go

```diff
@@ -76,62 +99,290 @@ func newBlockPrefetcher(retriever BlockRetriever,
+func (p *blockPrefetcher) shutdownLoop() {
+top:
+	for {
+		select {
+		case chInterface := <-p.inFlightFetches.Out():
+			ch := chInterface.(<-chan error)
+			<-ch
+		case <-p.shutdownCh:
+			break top
+		}
+	}
+	for p.inFlightFetches.Len() > 0 {
+		chInterface := <-p.inFlightFetches.Out()
+		ch := chInterface.(<-chan error)
+		<-ch
+	}
+	p.almostDoneCh <- struct{}{}
+}
+
+// run prefetches blocks.
+// E.g. a synced prefetch:
+// a -> {b -> {c, d}, e -> {f, g}}:
+// * state of prefetch tree in `p.prefetches`.
+// 1) a is fetched, triggers b and e.
+//    * a:2 -> {b:1, e:1}
+// 2) b is fetched, decrements b and a by 1, and triggers c and d to increment
+//    b and a by 2.
+//    * a:3 -> {b:2 -> {c:1, d:1}, e:1}
+// 3) c is fetched, and isTail==true so it completes up the tree.
+//    * a:2 -> {b:1 -> {d:1}, e:1}
+// 4) d is fetched, and isTail==true so it completes up the tree.
+//    * a:1 -> {e:1}
+// 5) e is fetched, decrements e and a by 1, and triggers f and g to increment
+//    e an a by 2.
+//    * a:2 -> {e:2 -> {f:1, g:1}}
+// 6) f is fetched, and isTail==true so it completes up the tree.
+//    * a:1 -> {e:1 -> {g:1}}
+// 5) g is fetched, completing g, e, and a.
+//    * <empty>
+//
+// A *critical* assumption here is that a block tree will never have a diamond
+// topology. That is, while a block may have multiple parents, at no point can
+// there exist more than one path from a block to another block in the tree.
+// That assumption should hold because blocks are content addressed, so
+// changing anything about one block creates brand new parents all the way up
+// the tree. If this did ever happen, a completed fetch downstream of the
+// diamond would be double counted in all nodes above the diamond, and the
+// prefetcher would eventually panic.
+func (p *blockPrefetcher) run(testSyncCh <-chan struct{}) {
 	defer func() {
-		wg.Wait()
 		close(p.doneCh)
 	}()
+	isShuttingDown := false
 	for {
+		if isShuttingDown && p.inFlightFetches.Len() == 0 &&
+			p.prefetchRequestCh.Len() == 0 && p.prefetchCancelCh.Len() == 0 {
+			return
+		}
+		if testSyncCh != nil {
+			<-testSyncCh
+		}
 		select {
-		case req := <-p.progressCh:
-			ctx, cancel := context.WithTimeout(context.Background(),
-				prefetchTimeout)
-			errCh := p.retriever.RequestWithPrefetch(ctx, req.priority,
-				req.kmd, req.ptr, req.block, TransientEntry, req.doneCh,
-				req.errCh)
-			wg.Add(1)
-			go func() {
-				defer wg.Done()
-				defer cancel()
-				select {
-				case err := <-errCh:
-					if err != nil {
-						p.log.CDebugf(ctx, "Done prefetch for block %s. "+
-							"Error: %+v", req.ptr.ID, err)
+		case bid := <-p.prefetchCancelCh.Out():
+			blockID := bid.(kbfsblock.ID)
+			pre, ok := p.prefetches[blockID]
+			if !ok {
+				continue
+			}
+			p.log.Debug("canceling prefetch for block %s", blockID)
+			// Walk up the block tree and delete every parent.
+			p.applyToParentsRecursive(p.cancelPrefetch, blockID, pre)
```

Comment:
The role of a canceled prefetch is to notify the parent tree that it can never get to a `FinishedPrefetch` state unless it does something. If the current block fails to prefetch for some reason, its parents need to find out. But its children still have a chance to finish prefetching and thus set their own prefetch state to `FinishedPrefetch`. This will save us from having to prefetch that child's tree later.

Reply:
Ah ok, kind of the reverse of my intuition.  Not sure how to make it more clear though.  Fine as is I guess.